SwiftCompilerSources/Sources/Optimizer/FunctionPasses/ReleaseDevirtualizer.swift

```diff
@@ -0,0 +1,201 @@
+  // in deallocating state, which means set the RC_DEALLOCATING_FLAG flag.
+  builder.createSetDeallocating(operand: object, isAtomic: release.isAtomic)
+
+  // Create the call to the destructor with the allocated object as self
+  // argument.
+  let functionRef = builder.createFunctionRef(dealloc)
+
+  let substitutionMap = context.getContextSubstitutionMap(for: type)
+  builder.createApply(function: functionRef, substitutionMap, arguments: [object])
+  context.erase(instruction: release)
+}
+
+private func stripRCIdentityPreservingInsts(_ value: Value) -> Value? {
+  guard let inst = value as? Instruction else { return nil }
+
+  switch inst {
+  // First strip off RC identity preserving casts.
+  case is UpcastInst, 
+    is UncheckedRefCastInst,
+    is InitExistentialRefInst,
+    is OpenExistentialRefInst,
+    is RefToBridgeObjectInst,
+    is BridgeObjectToRefInst,
+    is ConvertFunctionInst,
+    is UncheckedEnumDataInst:
+    return inst.operands[0].value
+
+  // Then if we have a struct_extract that is extracting a non-trivial member
+  // from a struct with no other non-trivial members, a ref count operation on
+  // the struct is equivalent to a ref count operation on the extracted
+  // member. Strip off the extract.
+  case let sei as StructExtractInst where sei.isFieldOnlyNonTrivialField:
+    return sei.operand
+
+  // If we have a struct or tuple instruction with only one non-trivial operand, the
+  // only reference count that can be modified is the non-trivial operand. Return
+  // the non-trivial operand.
+  case is StructInst, is TupleInst:
+    return inst.uniqueNonTrivialOperand
+
+  // If we have an enum instruction with a payload, strip off the enum to
+  // expose the enum's payload.
+  case let ei as EnumInst where !ei.operands.isEmpty:
+    return ei.operand
+
+  // If we have a tuple_extract that is extracting the only non trivial member
+  // of a tuple, a retain_value on the tuple is equivalent to a retain_value on
+  // the extracted value.
+  case let tei as TupleExtractInst where tei.isEltOnlyNonTrivialElt:
+    return tei.operand
+
+  default:
+    return nil
+  }
+}
+
+private extension Instruction {
+  /// Search the operands of this tuple for a unique non-trivial elt. If we find
+  /// it, return it. Otherwise return `nil`.
+  var uniqueNonTrivialOperand: Value? {
+    var candidateElt: Value?
+    let function = self.function
+
+    for op in operands {
+      if !op.value.type.isTrivial(in: function) {
+        if candidateElt == nil {
+          candidateElt = op.value
+          continue
+        }
+
+        // Otherwise, we have two values that are non-trivial. Bail.
+        return nil
+      }
+    }
+
+    return candidateElt
+  }
+}
+
+private extension TupleExtractInst {
+  var isEltOnlyNonTrivialElt: Bool {
+    let function = self.function
+
+    if type.isTrivial(in: function) {
+      return false
+    }
+
+    let opType = operand.type
+
+    return opType.tupleElements.filter { 
```

Comment:
Unfortunately `filter` is pretty bad for performance because it creates a new Array (involving memory allocation, etc.).
You can do `lazy.filter`, but then you cannot use `count`.
Or you do a good old fashioned `for ... in ...` loop (which would have the benefit that you abort as soon as the count exceeds one.